glfs.c

```diff
@@ -539,54 +550,122 @@ static void tcmu_glfs_close(struct tcmu_device *dev)
 	free(gfsp);
 }
 
-static int tcmu_glfs_read(struct tcmu_device *dev, struct tcmulib_cmd *cmd,
-			  struct iovec *iov, size_t iov_cnt, size_t length,
-			  off_t offset)
+static void glfs_async_cbk(glfs_fd_t *fd, ssize_t ret, void *data)
 {
-        struct glfs_state *state = tcmu_get_dev_private(dev);
-	ssize_t ret;
-
-        ret = glfs_preadv(state->gfd, iov, iov_cnt, offset, SEEK_SET);
-	if (ret != length) {
-		ret =  tcmu_set_sense_data(cmd->sense_buf, MEDIUM_ERROR,
-					   ASC_READ_ERROR, NULL);
+	glfs_cbk_cookie *cookie = data;
+	struct tcmu_device *dev = cookie->dev;
+	struct tcmulib_cmd *cmd = cookie->cmd;
+	size_t length = cookie->length;
+
+	if (ret < 0 || ret != length) {
+		/* Read/write/flush failed */
+		switch (cookie->op) {
+		case TCMU_GLFS_READ:
+			ret =  tcmu_set_sense_data(cmd->sense_buf, MEDIUM_ERROR,
+			                           ASC_READ_ERROR, NULL);
+			break;
+		case TCMU_GLFS_WRITE:
+		case TCMU_GLFS_FLUSH:
+			ret =  tcmu_set_sense_data(cmd->sense_buf, MEDIUM_ERROR,
+			                           ASC_WRITE_ERROR, NULL);
+			break;
+		}
 	} else {
 		ret = SAM_STAT_GOOD;
 	}
 
 	cmd->done(dev, cmd, ret);
+	free(cookie);
+}
+
+static int tcmu_glfs_read(struct tcmu_device *dev,
+                          struct tcmulib_cmd *cmd,
+                          struct iovec *iov, size_t iov_cnt,
+                          size_t length, off_t offset)
+{
+	struct glfs_state *state = tcmu_get_dev_private(dev);
+	glfs_cbk_cookie *cookie;
+
+	cookie = calloc(1, sizeof(*cookie));
+	if (!cookie) {
+		tcmu_err("Could not allocate cookie: %m\n");
+		goto out;
+	}
+	cookie->dev = dev;
+	cookie->cmd = cmd;
+	cookie->length = length;
+	cookie->op = TCMU_GLFS_READ;
+
+	if (glfs_preadv_async(state->gfd, iov, iov_cnt, offset, SEEK_SET,
```

Comment:
Will your async readv/writev  calls modify the iovec's iov_len or iov_base fields? For example for short read/writes would it have partially updated the iov_len field to indicated where it left off?

Just asking, because I am trying to track down what all the handlers do.